polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -55,18 +55,56 @@
 //!    older by a predefined parameter. This process is run very rarely (say, once a day). Once the
 //!    artifact is expired it is removed from disk eagerly atomically.
 
-use crate::host::PrepareResultSender;
+use crate::{host::PrepareResultSender, LOG_TARGET};
 use always_assert::always;
+use polkadot_core_primitives::Hash;
 use polkadot_node_core_pvf_common::{error::PrepareError, prepare::PrepareStats, pvf::PvfPrepData};
 use polkadot_node_primitives::NODE_VERSION;
 use polkadot_parachain_primitives::primitives::ValidationCodeHash;
 use polkadot_primitives::ExecutorParamsHash;
 use std::{
 	collections::HashMap,
 	path::{Path, PathBuf},
+	str::FromStr as _,
 	time::{Duration, SystemTime},
 };
 
+macro_rules! concat_const {
```

Comment:
Now that I think of it, I could probably generalize the macro.

Reply:
> emit some compile-time error instead, that would be ideal

Turns out panic is const. Since const evalutation happens during compilation, so this is practically a compile-time error (as long as we don't export the macro).

```rs
match std::str::from_utf8(&CAT) {
    Ok(s) => s,
    Err(err) => panic!("error converting bytes to str"),
}
```

If unsafe if the main concern here, does this make the macro ok? :)

But again, the unchecked version should be safe in the first place, as `&str` share the same memory layout with those bytes slice.

Reply:
* concationation of two `str`s is guaranteed to be valid utf8
* `s` and `s.as_bytes()` share exactly the same memory layout

So the concatentation of the byte slices is also always valid. If not, that's our fault. ;) It's this kind of guarantees make Rust so lovable.

Reply:
> concationation of two strs is guaranteed to be valid utf8

I didn't know that! Where is it documented? We should document it in the macro.

Reply:
https://github.com/rossmacarthur/constcat/blob/trunk/src/lib.rs

Just found out about this crate that's basically doing the same thing as this macro. They use `from_utf8_unchecked` for the same safety reasoning.

Reply:
Well, just because someone made a crate for it, doesn't make it sound. 😛

[The repo](https://github.com/rossmacarthur/constcat) only has 5 stars, which doesn't inspire confidence. The SAFETY comment there is also missing a step in the reasoning:

        // SAFETY: The original constants were asserted to be &str's
        // so the resultant bytes are valid UTF-8.

I know I sound like a broken record, but where is it documented that directly concatenating `str` bytes results in valid UTF-8? I'm not saying it's unsound, just that we need stronger proofs when introducing `unsafe`.

Reply:
Good to know, thanks 💪